### PR TITLE
Kallisto package and Trinity integration

### PR DESCRIPTION
**Pull request by abretaud**
Here is the part from #602 concerning the integration of kallisto with trinity.
Waiting for an answer from kallisto authors about the licence problem

### CODE REVIEW

**Comment by abretaud** [comment]
It looks like the authors are ok: https://groups.google.com/forum/#!searchin/kallisto-sleuth-users/galaxy/kallisto-sleuth-users/oVRg69Zmots/toixrB3SDAAJ
I think this PR should stay in standby mode until #602 is merged

File: tools/trinity/align_and_estimate_abundance.xml [hunk 1]
```diff
@@ -0,0 +1,292 @@
...
+        <requirement type="package" version="1.2">samtools</requirement>
+        <requirement type="package" version="1.2.28">rsem</requirement>
+        <requirement type="package" version="0.42.4">kallisto</requirement>
+        <requirement type="package" version="1.5.1">express</requirement>
+    </requirements>
+    <stdio>
+        <exit_code range="1:"/>
+    </stdio>
+    <command><![CDATA[
+        ln -s $transcripts input.fa
```
Comment (yhoogstrate) [inline tools/trinity/align_and_estimate_abundance.xml]: `$transcripts`

File: tools/trinity/align_and_estimate_abundance.xml [hunk 2]
```diff
@@ -0,0 +1,292 @@
+<tool id="align_and_estimate_abundance" name="Align reads and estimate abundance" version="2.1.1">
+    <description>on a de novo assembly of RNA-Seq data</description>
+    <requirements>
+        <requirement type="package" version="2.1.1">trinity</requirement>
+        <requirement type="package" version="1.1.2">bowtie</requirement>
+        <requirement type="package" version="2.2.6">bowtie2</requirement>
+        <requirement type="package" version="1.2">samtools</requirement>
+        <requirement type="package" version="1.2.28">rsem</requirement>
+        <requirement type="package" version="0.42.4">kallisto</requirement>
+        <requirement type="package" version="1.5.1">express</requirement>
```
Comment (yhoogstrate) [inline tools/trinity/align_and_estimate_abundance.xml]: perl?

File: tools/trinity/align_and_estimate_abundance.xml [hunk 3]
```diff
@@ -0,0 +1,292 @@
...
+            #else:
+                --seqType fq
+            #end if
+            
+            #if $inputs.strand.is_strand_specific:
+                --SS_lib_type $inputs.strand.library_type
+            #end if
+
+        #else:
+            --single $inputs.input
```
Comment (yhoogstrate) [inline tools/trinity/align_and_estimate_abundance.xml]: quotes

File: tools/trinity/align_and_estimate_abundance.xml [hunk 4]
```diff
@@ -0,0 +1,292 @@
...
+            #if $inputs.strand.is_strand_specific:
+                --SS_lib_type $inputs.strand.library_type
+            #end if
+        #end if
+        
+        --max_ins_size $inputs.paired_fragment_length
+
+        ## Additional parameters.
+        #if $additional_params.gene_map.has_gene_map == "yes":
+            --gene_trans_map $additional_params.gene_map.gene_trans_map
```
Comment (yhoogstrate) [inline tools/trinity/align_and_estimate_abundance.xml]: `'...'`

File: tools/trinity/align_and_estimate_abundance.xml [hunk 5]
```diff
@@ -0,0 +1,292 @@
...
+            </when>
+            <when value="single">
+                <param format="fasta,fastqsanger" name="input" type="data" label="Single-end reads" help=""/>
+                <conditional name="strand">
+                    <param name="is_strand_specific" type="boolean" checked="false" label="Strand specific data"/>
+                    <when value="false">
+                    </when>
+                    <when value="true">
+                        <param name="library_type" type="select" label="Strand-specific Library Type">
+                            <option value="F">F</option>
```
Comment (yhoogstrate) [inline tools/trinity/align_and_estimate_abundance.xml]: Could the display value be a more informative?

File: tools/trinity/align_and_estimate_abundance.xml [hunk 6]
```diff
@@ -0,0 +1,292 @@
...
+        </conditional>
+        
+        <conditional name="method">
+            <param type="select" name="est_method" label="Abundance estimation method">
+                <option value="RSEM">RSEM</option>
+                <option value="eXpress">eXpress</option>
+                <option value="kallisto">kallisto</option>
+            </param>
+            <when value="RSEM">
+                <param type="select" name="aln_method" label="Alignment method">
```
Comment (yhoogstrate) [inline tools/trinity/align_and_estimate_abundance.xml]: Could you put this in a macro and copy it also to down below?

File: tools/trinity/align_and_estimate_abundance.xml [hunk 7]
```diff
@@ -0,0 +1,292 @@
...
+                </when>
+                <when value="yes">
+                    <param format="tabular" name="gene_trans_map" type="data" label="Gene to transcript correspondence ('gene(tab)transcript' lines)" />
+                </when>
+            </conditional>
+            
+        </section>
+    </inputs>
+    <outputs>
+        <data format="tabular" name="isoforms_counts_rsem" label="${tool.name} on ${on_string}: isoforms counts" from_work_dir="RSEM.isoforms.results">
```
Comment (yhoogstrate) [inline tools/trinity/align_and_estimate_abundance.xml]: how about: isform counts (RSEM)

File: tools/trinity/align_and_estimate_abundance.xml [hunk 8]
```diff
@@ -0,0 +1,292 @@
...
+                            <option value="F">F</option>
+                            <option value="R">R</option>
+                        </param>
+                    </when>
+                </conditional>
+            </when>
+        </conditional>
+        
+        <conditional name="method">
+            <param type="select" name="est_method" label="Abundance estimation method">
```
Comment (yhoogstrate) [inline tools/trinity/align_and_estimate_abundance.xml]: I don't know if this is possible with trinity and if it makes sense to you, but would it not be more convenient to have 3 checkboxes:

*Choose output:*
 - [ ] RSEM
 - [ ] eXpress
 - [ ] Kallisto

such that all three output files can be generated in one run?

File: tools/trinity/align_and_estimate_abundance.xml [hunk 9]
```diff
@@ -0,0 +1,292 @@
...
+
+.. _Trinity manual: https://github.com/trinityrnaseq/trinityrnaseq/wiki/Trinity-Transcript-Quantification
+
+
+.. _Trinity: http://trinityrnaseq.github.io
+]]>
+    </help>
+    
+     <citations>
+        <citation type="doi">doi:10.1038/nbt.1883</citation>
```
Comment (yhoogstrate) [inline tools/trinity/align_and_estimate_abundance.xml]: `<citation type="doi">doi:10.1038/nbt.1883</citation>` like (`sed 's/doi://g'`)

**Comment by abretaud** [comment]
Thank you for reviewing @yhoogstrate!
I had forgotten about this PR... I think it would need to be upgraded as there were some changes in trinity wrappers.
*But* as there is a problem on the kallisto licence, and as there is now another mapping-free method (salmon) available, I'm wondering if we'd better just close this?

**Comment by yhoogstrate** [comment]
I discussed this with @bgruening and you're right; the authors of kallisto do not want their software to be used nor (re)distributed for any other reason than education and non-profitable research. This leaves us no option than to stop here :-1: . Thanks for the efford anyway @abretaud :+1:

**Comment by wilzbach** [comment]
> The authors of kallisto do not want their software to be used nor (re)distributed for any other reason than education and non-profitable research.

@ncbray @pimentel @pmelsted @lakigigar
Seriously?? What happened to open research for the good of humanity?
Apart from being really disappointed, I would like to point out that you copy & pasted some [MIT software](https://github.com/pachterlab/kallisto/blob/master/src/kseq.h#L1) in your source code, which you have to at least mention in your copyright notice. Moreover do you really have the license for the [Ursa major image](http://fineartamerica.com/featured/ursa-major-constellation-hevelius-1687-science-source.html) which you reproduced on your website?
I hope you are aware that it was only possible for you to use & modify all the libraries and tools in your work, because other people weren't as selfish as you?

**Comment by peterjc** [comment]
@wilzbach wrote:

> Seriously?? What happened to open research for the good of humanity?

There isn't currently an explicit code of conduct on the tools-iuc repository, but if there was it would likely match that of the main Galaxy repository. With that in mind, I would encourage you try to be more considerate/respectful in your phrasing - see: https://github.com/galaxyproject/galaxy/blob/dev/CODE_OF_CONDUCT.md

I do however share your frustration with the non-open source licensing of Kallisto. This puts up hurdles for automated packages and deployment, and is problematic for semi-commercial research like industry partnerships (e.g. part Industry funded PhD positions, or Industry commissioned research done at a non-profit Institute).

**Comment by martenson** [comment]
Thanks to the license change, new version of kallisto is now wrapped. https://github.com/galaxyproject/tools-iuc/pull/1427

**Comment by nsoranzo** [comment]
Maybe @abretaud is still interested in adding kallisto support to `tools/trinity/abundance_estimates_to_matrix.xml` and `tools/trinity/align_and_estimate_abundance.xml` ?

**Comment by abretaud** [comment]
Great! I'll work on it, I'm reopening it to keep it in mind

**Comment by abretaud** [comment]
Ok, it's green, ready for a little review if someone is available!
I don't really understand why this time I had to add the `r-cluster` dependency to some other tools, I guess we should have seen it on previous PR...

**Comment by bgruening** [comment]
This looks very good! :+1:

**Comment by martenson** [comment]
💯

**Pull request merged**